### PR TITLE
Let `Radix2Domain::offset` to be `FpVar` instead of `F`

### PR DESCRIPTION
**Pull request by tsunrise**
## Description

Current `Radix2Domain` does not support non-constant offset. This PR adds that functionality. 

**Highlighted Changes:** 
- `Radix2Domain` will be renamed to `Radix2DomainVar`
- `Radix2DomainVar` will offset that is `FpVar` instead of `F`
- `Radix2DomainVar::interpolate_and_evaluate` will specialize according to whether `self.offset` is constant or not. 

closes: #64 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by weikengchen** [comment]
The change looks good. Although this is a breaking change, the Radix2Domain in r1cs_std is not used very often downstream (most are using the Radix2Domain in ark-ff), and I think renaming makes sense.

File: src/poly/domain/mod.rs [hunk 1]
```diff
@@ -1,29 +1,40 @@
...
+    pub offset: FpVar<F>,
     /// dimension of evaluation domain
     pub dim: u64,
 }
 
-impl<F: PrimeField> Radix2Domain<F> {
+impl<F: PrimeField> Radix2DomainVar<F> {
+    /// returns whether two domains are same without comparing if `offset`
+    /// represents same variable in circuit.
+    pub fn is_same_value(&self, other: &Self) -> bool {
```
Comment (Pratyush) [inline src/poly/domain/mod.rs:30]: Why is this necessary?

File: src/poly/domain/mod.rs [hunk 2]
```diff
@@ -40,6 +51,11 @@ impl<F: PrimeField> Radix2Domain<F> {
         result
     }
 
+    /// Size of coset
+    pub fn size(&self) -> u64 {
+        1 << self.dim
+    }
```
Comment (Pratyush) [inline src/poly/domain/mod.rs:57]:
```suggestion
    /// Size of the domain.
    pub fn size(&self) -> u64 {
        1 << self.dim
    }
```

File: src/poly/evaluations/univariate/mod.rs [hunk 3]
```diff
@@ -18,7 +18,12 @@ pub struct EvaluationsVar<F: PrimeField> {
     pub evals: Vec<FpVar<F>>,
     /// Optional Lagrange Interpolator. Useful for lagrange interpolation.
     pub lagrange_interpolator: Option<LagrangeInterpolator<F>>,
-    domain: Radix2Domain<F>,
+    domain: Radix2DomainVar<F>,
+    /// Contains all domain elements of `domain.base_domain`.
+    ///
+    /// This is Cache for lagrange interpolation when offset is non-constant. Will be `None` if offset is constant
```
Comment (Pratyush) [inline src/poly/evaluations/univariate/mod.rs:24]:
```suggestion
    /// This is a cache for lagrange interpolation when offset is non-constant. Will be `None` if offset is constant
```

File: src/poly/domain/mod.rs (hunk 1 above)
Comment (tsunrise) [reply src/poly/domain/mod.rs:30]: Is there any case when we wantsto check equality of domain? If so we can change it to `is_eq(&self, other: &Self) -> Result<Boolean<F>, SynthesisError>`

File: src/poly/domain/mod.rs (hunk 2 above)
Comment (tsunrise) [reply src/poly/domain/mod.rs]: ack

File: src/poly/evaluations/univariate/mod.rs (hunk 3 above)
Comment (tsunrise) [reply src/poly/evaluations/univariate/mod.rs]: ack

**Review by tsunrise — COMMENTED** [review]
Thanks for review! fixed 2 of 3

File: src/poly/domain/mod.rs (hunk 1 above)
Comment (weikengchen) [reply src/poly/domain/mod.rs:30]: maybe `EqGadget` makes more sense

Comment (weikengchen) [reply src/poly/domain/mod.rs:30]: also, why it is "without comparing if `offset` represents same variable in circuit.", but the `offset` is indeed considered?

Comment (tsunrise) [reply src/poly/domain/mod.rs]: "without comparing if offset represents same variable in circuit." seems problematic. Anyway, I just pushed a commit to replace it with `EqGadget` instead

**Comment by weikengchen** [comment]
will wait for @ValarDragon for a final review

File: src/poly/domain/mod.rs [hunk 4]
```diff
@@ -7,23 +8,33 @@ use ark_std::vec::Vec;
...
 /// Defines an evaluation domain over a prime field. The domain is a coset of size `1<<dim`.
 ///
 /// Native code corresponds to `ark-poly::univariate::domain::radix2`, but `ark-poly` only supports
 /// subgroup for now.
 ///
 /// TODO: support cosets in `ark-poly`.
-pub struct Radix2Domain<F: PrimeField> {
+pub struct Radix2DomainVar<F: PrimeField> {
     /// generator of subgroup g
     pub gen: F,
```
Comment (ValarDragon) [inline src/poly/domain/mod.rs:20]: Should we plan to make this FpVar<F> in a future release as well?

File: src/poly/evaluations/univariate/mod.rs [hunk 5]
```diff
@@ -40,22 +45,39 @@ impl<F: PrimeField> EvaluationsVar<F> {
...
 
-    /// Generate lagrange interpolator and mark it ready to interpolate
-    pub fn generate_lagrange_interpolator(&mut self) {
-        let poly_evaluations_val: Vec<_> = self.evals.iter().map(|v| v.value().unwrap()).collect();
-        let domain = &self.domain;
-        let lagrange_interpolator =
-            LagrangeInterpolator::new(domain.offset, domain.gen, domain.dim, poly_evaluations_val);
-        self.lagrange_interpolator = Some(lagrange_interpolator)
+    /// Precompute necessary calculation for lagrange interpolation and mark it ready to interpolate
+    pub fn generate_interpolate_cache(&mut self) {
```
Comment (ValarDragon) [inline src/poly/evaluations/univariate/mod.rs:57]:
```suggestion
    pub fn generate_interpolation_cache(&mut self) {
```

**Review by ValarDragon — APPROVED** [review]
LGTM, just a suggestion for one minor function rename

File: src/poly/domain/mod.rs (hunk 4 above)
Comment (weikengchen) [reply src/poly/domain/mod.rs:20]: I did some preliminary attempts. It is possible but would require a lot of changes. I guess we would leave it to the next release.

There are some potential use cases. For example, a Marlin gadget verifies proofs of varying domain sizes, e.g., if used as layer-2 of ZEXE.

Comment (tsunrise) [reply src/poly/domain/mod.rs:20]: Sounds good. I will put this on my stack, and will probably do so after I finish up FRI protocol
